src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java

```diff
@@ -1,25 +1,79 @@
 package com.es.phoneshop.model.product;
 
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Currency;
 import java.util.List;
+import java.util.stream.Collectors;
 
 public class ArrayListProductDao implements ProductDao {
+    private long maxId;
+    private List<Product> products;
+
+    public ArrayListProductDao() {
+        this.products = new ArrayList<>();
+        saveSampleProducts();
+    }
+
     @Override
-    public Product getProduct(Long id) {
-        throw new RuntimeException("Not implemented");
+    public synchronized Product getProduct(Long id) throws ProductNotFoundException {
+        return products.stream()
+                .filter(product -> id.equals(product.getId()))
+                .findAny()
+                .orElseThrow(ProductNotFoundException::new);
     }
 
     @Override
-    public List<Product> findProducts() {
-        throw new RuntimeException("Not implemented");
+    public synchronized List<Product> findProducts() {
+        return products.stream()
+                .filter(product -> product.getPrice() != null)
+                .filter(product -> product.getStock() > 0)
+                .collect(Collectors.toList());
     }
 
     @Override
-    public void save(Product product) {
-        throw new RuntimeException("Not implemented");
+    public synchronized void save(Product product) {
+        if (product.getId() != null) {
+            Long id = product.getId();
+            if (products.stream().noneMatch(p -> id.equals(p.getId()))) {
+                products.add(product);
+            } else {
+                Product oldProduct = products.stream()
+                        .filter(p -> id.equals(p.getId()))
+                        .findAny()
+                        .get();
+                Collections.replaceAll(products, oldProduct, product);
```

Comment:
Do you really need to traverse products list twice here? It's working but I think it can be more effective.

Reply:
Fixed